### PR TITLE
Base toss

### PR DESCRIPTION
**Pull request by liuyu85cn**
This PR is to address this issue that in-edge and out-edge may in-consist while add and update.

I add an upper tier above the normal AddEdgeProcess / UpdateEdgeProcessor.

and I named them Chain...Processor.

spaces who set TOSS on will call those chain processor, and the chain processor will drive the normal ones

to do the actually insert/update.

### CODE REVIEW

File: src/clients/meta/MetaClient.h [hunk 1]
```diff
@@ -24,6 +24,14 @@
 
 DECLARE_int32(meta_client_retry_times);
 
+namespace nebula {
+namespace storage {
+class ExtMetaClientModifier;
+class MetaClientTestUpdater;
+class ChainAddEdgesTest_AddEdgesLocalTest_Test;
+}  // namespace storage
+}  // namespace nebula
```
Comment (yixinglu) [inline src/clients/meta/MetaClient.h:33]: It seems these forward declarations are not necessary and I don't understand why to make the MetaClient to be the friend of ChainAddEdges?

File: src/storage/transaction/ChainAddEdgesProcessorLocal.h [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+#include "interface/gen-cpp2/storage_types.h"
+#include "kvstore/LogEncoder.h"
+#include "storage/BaseProcessor.h"
+#include "storage/transaction/ConsistUtil.h"
+#include "storage/transaction/TransactionManager.h"
+
+namespace nebula {
+namespace storage {
+
+class ChainAddEdgesProcessorLocal : public BaseProcessor<cpp2::ExecResponse>,
```
Comment (yixinglu) [inline src/storage/transaction/ChainAddEdgesProcessorLocal.h:18]: LocalChainAddEdgesProcessor ?

**Comment by critical27** [comment]
Close #2617

File: src/storage/transaction/ChainAddEdgesProcessorLocal.cpp [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+  auto isrc = toInt(edge.get_key().get_src().toString());
+  auto idst = toInt(edge.get_key().get_dst().toString());
+  std::stringstream oss;
+  oss << isrc << "->" << idst << ", val: ";
+  for (auto& val : edge.get_props()) {
+    oss << val.toString() << " ";
+  }
+  return oss.str();
+}
+
```
Comment (bright-starry-sky) [inline src/storage/transaction/ChainAddEdgesProcessorLocal.cpp:57]: It's too useful.

Comment (bright-starry-sky) [inline src/storage/transaction/ChainAddEdgesProcessorLocal.cpp:57]: It's too useful.

File: src/storage/transaction/ChainAddEdgesProcessorLocal.cpp [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+  }
+
+  if (!lockEdges(req_)) {
+    return Code::E_WRITE_WRITE_CONFLICT;
+  }
+
+  auto [pro, fut] = folly::makePromiseContract<Code>();
+  auto primes = makePrime();
+  for (auto& kv : primes) {
+    LOG(INFO) << uuid_ << " put prime " << folly::hexlify(kv.first);
```
Comment (bright-starry-sky) [inline src/storage/transaction/ChainAddEdgesProcessorLocal.cpp:78]: what's about VLOG ?

**Review by panda-sheep — COMMENTED** [review]
A impressive job, which has made our products a big step forward. 👍

File: src/clients/storage/GraphStorageClient.cpp [hunk 5]
```diff
@@ -165,7 +165,7 @@ folly::SemiFuture<StorageRpcResponse<cpp2::ExecResponse>> GraphStorageClient::ad
       evb,
       std::move(requests),
       [=](cpp2::GraphStorageServiceAsyncClient* client, const cpp2::AddEdgesRequest& r) {
-        return useToss ? client->future_addEdgesAtomic(r) : client->future_addEdges(r);
+        return useToss ? client->future_chainAddEdges(r) : client->future_addEdges(r);
```
Comment (panda-sheep) [inline src/clients/storage/GraphStorageClient.cpp:168]: In `src/storage/InternalStorageServiceHandler.h ` file, the `future_chainAddEdges ` should be InterStorageServiceAsyncClient?

Comment (panda-sheep) [reply src/clients/storage/GraphStorageClient.cpp:168]: Please ignore this, both `InternalStorageService` and `GraphStorageService` have this function.

File: src/storage/transaction/ChainAddEdgesProcessorLocal.cpp (hunk 4 above)
Comment (liuyu85cn) [reply src/storage/transaction/ChainAddEdgesProcessorLocal.cpp]: good catch

File: src/clients/storage/StorageClientBase.h [hunk 6]
```diff
@@ -209,13 +209,13 @@ class StorageClientBase {
     return {req.get_part_id()};
   }
 
-  std::vector<PartitionID> getReqPartsId(const cpp2::InternalTxnRequest& req) const {
-    return {req.get_part_id()};
-  }
+  //   std::vector<PartitionID> getReqPartsId(const cpp2::InternalTxnRequest& req) const {
```
Comment (SuperYoko) [inline src/clients/storage/StorageClientBase.h:212]: remove unused code.

File: src/storage/transaction/TransactionManager.cpp [hunk 7]
```diff
@@ -8,414 +8,195 @@
...
-                << ", txnId=" << txnId;
-    }
+TransactionManager::LockCore* TransactionManager::getLockCore(GraphSpaceID spaceId,
+                                                              GraphSpaceID partId) {
+  if (whiteListParts_.find(std::make_pair(spaceId, partId)) == whiteListParts_.end()) {
+    LOG(WARNING) << folly::sformat("space {}, part{} not in white list");
+    scanPrimes(spaceId, partId);
+    auto key = std::make_pair(spaceId, partId);
+    whiteListParts_.insert(std::make_pair(key, 0));
+    return nullptr;
```
Comment (critical27) [inline src/storage/transaction/TransactionManager.cpp:42]: Pls check the logic around whiteListParts_

File: src/meta/processors/parts/GetPartsAllocProcessor.cpp [hunk 8]
```diff
@@ -34,8 +34,60 @@ void GetPartsAllocProcessor::process(const cpp2::GetPartsAllocReq& req) {
...
+
+  std::vector<PartitionID> partIdVec;
+  std::vector<std::string> leaderKeys;
+  for (auto partId = 1; partId <= partNum; ++partId) {
+    partIdVec.emplace_back(partId);
+    leaderKeys.emplace_back(MetaServiceUtils::leaderKey(spaceId, partId));
+  }
+
+  std::vector<std::string> vals;
+  auto [code, statusVec] = kvstore_->multiGet(0, 0, std::move(leaderKeys), &vals);
```
Comment (panda-sheep) [inline src/meta/processors/parts/GetPartsAllocProcessor.cpp:68]: please `kDefaultSpaceId`, `kDefaultPartId`

File: src/meta/processors/parts/GetPartsAllocProcessor.cpp [hunk 9]
```diff
@@ -34,8 +34,60 @@ void GetPartsAllocProcessor::process(const cpp2::GetPartsAllocReq& req) {
...
+  if (code != nebula::cpp2::ErrorCode::SUCCEEDED &&
+      code != nebula::cpp2::ErrorCode::E_PARTIAL_RESULT) {
+    LOG(INFO) << "error rc = " << apache::thrift::util::enumNameSafe(code);
+    return ret;
```
Comment (panda-sheep) [inline src/meta/processors/parts/GetPartsAllocProcessor.cpp:72]: The `handleErrorCode `is missing here

This function returns `ErrorOr<nebula::cpp2::ErrorCode, std::unordered_map<PartitionID, TermID>>` is more appropriate.
In this way, handleErrorCode is handled uniformly outside

File: src/storage/transaction/TransactionManager.cpp (hunk 7 above)
Comment (liuyu85cn) [reply src/storage/transaction/TransactionManager.cpp]: done

File: src/interface/storage.thrift [hunk 10]
```diff
@@ -864,27 +866,49 @@ service GeneralStorageService {
...
 }
 
-struct GetValueRequest {
-    1: common.GraphSpaceID space_id,
-    2: common.PartitionID part_id,
-    3: binary key
+//
+// Response for data modification requests
+//
+struct ChainResponse {
```
[the remark this answers is not in the document]
Comment (liuyu85cn) [reply src/interface/storage.thrift:880]: good catch, removed

File: src/meta/ActiveHostsMan.cpp [hunk 11]
```diff
@@ -61,6 +60,8 @@ nebula::cpp2::ErrorCode ActiveHostsMan::updateHostInfo(kvstore::KVStore* kv,
       auto val = MetaServiceUtils::leaderValV3(hostAddr, terms[i]);
       data.emplace_back(std::make_pair(leaderKeys[i], std::move(val)));
     }
+  } else {
+    // do nothing
```
[the remark this answers is not in the document]
Comment (liuyu85cn) [reply src/meta/ActiveHostsMan.cpp:64]: ok

File: src/meta/processors/parts/GetPartsAllocProcessor.cpp (hunk 8 above)
Comment (liuyu85cn) [reply src/meta/processors/parts/GetPartsAllocProcessor.cpp:68]: ok

File: src/meta/processors/parts/GetPartsAllocProcessor.cpp (hunk 9 above)
Comment (liuyu85cn) [reply src/meta/processors/parts/GetPartsAllocProcessor.cpp:72]: for now, I consider term is an optional resp for GetPart. So not get a term will not set the GetPart RPC failed.

File: src/clients/storage/StorageClientBase.h (hunk 6 above)
Comment (liuyu85cn) [reply src/clients/storage/StorageClientBase.h:212]: ok

File: src/meta/processors/parts/GetPartsAllocProcessor.cpp (hunk 9 above)
Comment (panda-sheep) [reply src/meta/processors/parts/GetPartsAllocProcessor.cpp:73]: ok

File: src/storage/transaction/ChainAddEdgesProcessorRemote.cpp [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+
+void ChainAddEdgesProcessorRemote::forwardRequest(const cpp2::ChainAddEdgesRequest& req) {
+  auto spaceId = req.get_space_id();
+  auto* proc = AddEdgesProcessor::instance(env_);
+  proc->getFuture().thenValue([=](auto&& resp) {
+    Code rc = Code::SUCCEEDED;
+    for (auto& part : resp.get_result().get_failed_parts()) {
+      handleErrorCode(part.code, spaceId, part.get_part_id());
+    }
+    LOG(INFO) << this << " " << apache::thrift::util::enumNameSafe(rc);
```
Comment (bright-starry-sky) [inline src/storage/transaction/ChainAddEdgesProcessorRemote.cpp:58]: black technology

Comment (liuyu85cn) [reply src/storage/transaction/ChainAddEdgesProcessorRemote.cpp:58]: debug code... will change to vlog

File: src/storage/transaction/ConsistUtil.cpp [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+#include <boost/uuid/uuid.hpp>
+#include <boost/uuid/uuid_generators.hpp>
+#include <boost/uuid/uuid_io.hpp>
+
+#include "common/utils/NebulaKeyUtils.h"
+namespace nebula {
+namespace storage {
+
+static const std::string kPrimeTable{"__prime__"};              // NOLINT
+static const std::string kDoublePrimeTable{"__prime_prime__"};  // NOLINT
```
Comment (panda-sheep) [inline src/storage/transaction/ConsistUtil.cpp:18]: Can you add some notes about the system table?  thx

File: src/interface/common.thrift [hunk 14]
```diff
@@ -398,6 +398,11 @@ enum ErrorCode {
     E_TASK_EXECUTION_FAILED           = -3053,
 
     E_PLAN_IS_KILLED                  = -3060,
+    // toss
+    E_NO_TERM                         = -3070
```
Comment (darionyaphet) [inline src/interface/common.thrift:402]: maybe should support comma at the end

File: src/kvstore/test/RocksEngineTest.cpp [hunk 15]
```diff
@@ -8,6 +8,7 @@
 #include <gtest/gtest.h>
 #include <rocksdb/db.h>
 #include <rocksdb/table.h>
+#include <thrift/lib/cpp/util/EnumUtils.h>
```
Comment (darionyaphet) [inline src/kvstore/test/RocksEngineTest.cpp:11]: why ?

File: src/storage/transaction/ChainAddEdgesProcessorLocal.cpp [hunk 16]
```diff
@@ -0,0 +1,467 @@
...
+  req_ = req;
+  spaceId_ = req_.get_space_id();
+  auto vidType = env_->metaClient_->getSpaceVidType(spaceId_);
+  if (!vidType.ok()) {
+    LOG(WARNING) << "can't get vidType";
+    return false;
+  } else {
+    spaceVidType_ = vidType.value();
+  }
+  localPartId_ = req.get_parts().begin()->first;
```
[the remark this answers is not in the document]
Comment (liuyu85cn) [reply src/storage/transaction/ChainAddEdgesProcessorLocal.cpp:132]: yes, ChainAddEdgesProcessorLocal will only deal with one part (and one remote part).

**Comment by panda-sheep** [comment]
image

**Review by panda-sheep — APPROVED** [review]
Orz, brilliant!

**Review by critical27 — DISMISSED** [review]
Good job. Let's finish the long term run……

**Review by darionyaphet — DISMISSED** [review]

**Review by critical27 — APPROVED** [review]

**Review by bright-starry-sky — DISMISSED** [review]

**Review by panda-sheep — DISMISSED** [review]

**Comment by bright-starry-sky** [comment]
detected memory leaks

**Review by darionyaphet — APPROVED** [review]
LGTM

**Review by bright-starry-sky — APPROVED** [review]

**Review by panda-sheep — DISMISSED** [review]

**Review by darionyaphet — APPROVED** [review]

**Review by panda-sheep — APPROVED** [review]

**Pull request merged**